### ISSUE TITLE
[x86] convert xor with MIN_SIGNED_VALUE to add

### ISSUE THREAD

**Issue by rotateright** [llvm/llvm-project#51609]
|  |  |
| --- | --- |
| Bugzilla Link | [52267](https://llvm.org/bz52267) |
| Version | trunk |
| OS | All |
| CC | @topperc,@RKSimon,@phoebewang,@rotateright |

## Extended Description 
Based on discussion in https://reviews.llvm.org/D112085 :
```
define i32 @add_smin(i32 %x) {
  %r = add i32 %x, 2147483648
  ret i32 %r
}

define i32 @xor_smin(i32 %x) {
  %r = xor i32 %x, 2147483648
  ret i32 %r
}
```
These are logically equivalent:
https://alive2.llvm.org/ce/z/qV46E2

...and instcombine chooses 'xor' for better analysis in IR. But x86 (probably unlike most targets), may benefit from converting it back to 'add' (for legal integer types) because that can be lowered as LEA:
```
  movl	%edi, %eax
  xorl	$-2147483648, %eax              ## imm = 0x80000000
```
vs.
```
  leal	-2147483648(%rdi), %eax
```

**Comment by RKSimon** [comment]
Before adding X86 specific handling, I've proposed adding generic AddLike support in https://reviews.llvm.org/D122754

**Comment by RKSimon** [comment]
matchAddressRecursively support - https://reviews.llvm.org/D122815

**Comment by RKSimon** [comment]
> matchAddressRecursively support - https://reviews.llvm.org/D122815

Landed at c64f37f818b7a928b48540920d6e0c6df8903128

**Comment by RKSimon** [comment]
[X86] Add XOR(X, MIN_SIGNED_VALUE) -> ADD(X, MIN_SIGNED_VALUE) isel patterns (PR52267)
https://reviews.llvm.org/D123043

Landed at ffe0cc82dbaecde4fd1bb2bca5c6aea59be698a6

**Comment by RKSimon** [comment]
> Before adding X86 specific handling, I've proposed adding generic AddLike support in https://reviews.llvm.org/D122754

Landed at 3369e474bbd1172175945437783b6841979e654a
